File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>
using namespace std;

int n;
string s;
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    cin >> n;
    for(int i = 0; i < n; i++) {
        cin >> s;
        stack<char> st;
        bool flag = true;
        for(int j = 0; j < s.size(); j++) {
            if(s[j] == '(') {
                st.push(s[j]);
            }
            else if(st.empty() && s[j] == ')') {
                flag = false;
            }
            else if(st.size() && s[j] == ')' && st.top() == '(') {
                st.pop();
            }
        }
        if(st.empty() && flag) {
            cout << "YES" << '\n';
        }
        else {
            cout << "NO" << '\n';
        }
    }
    return 0;
}

File: 백준/Silver/9012. 괄호/괄호.java
import java.util.*;

public class Main {

    static int t;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        t = sc.nextInt();
        while(t > 0) {
            t--;
            Stack<Character> st = new Stack<>();
            String parenthesis = sc.next();
            boolean flag = true;
            for(int i = 0; i < parenthesis.length(); i++) {
                if(parenthesis.charAt(i) == '(') {
                    st.push(parenthesis.charAt(i));
                }
                else {
                    if(!st.isEmpty() && st.peek() == '(') {
                        st.pop();
                    }
                    else {
                        flag = false;
                        break;
                    }
                }
            }
            if(!st.isEmpty()) {
                flag = false;
            }
            if(flag) {
                System.out.println("YES");
            }
            else {
                System.out.println("NO");
            }
        }
    }
}
